offer-manager-domain/src/main/groovy/life/qbic/business/products/copy/CopyProduct.groovy

```diff
@@ -0,0 +1,134 @@
+package life.qbic.business.products.copy
+
+import life.qbic.business.Constants
+import life.qbic.business.exceptions.DatabaseQueryException
+import life.qbic.business.logging.Logger
+import life.qbic.business.logging.Logging
+import life.qbic.business.products.Converter
+import life.qbic.business.products.create.CreateProduct
+import life.qbic.business.products.create.CreateProductDataSource
+import life.qbic.business.products.create.CreateProductInput
+import life.qbic.business.products.create.CreateProductOutput
+import life.qbic.datamodel.dtos.business.services.Product
+import org.aspectj.bridge.IMessage
+
+import java.nio.charset.StandardCharsets
+import java.security.MessageDigest
+
+/**
+ * <h1>4.3.2 Copy Service Product</h1>
+ * <br>
+ * <p> Offer Administrators are allowed to create a new permutation of an existing product.
+ * <br> New permutations can include changes in unit price, sequencing technology and other attributes of service products.
+ * </p>
+ *
+ * @since: 1.0.0
+ *
+ */
+class CopyProduct implements CopyProductInput, CreateProductOutput {
+
+    private static final Logging log = Logger.getLogger(this.class)
+
+    private final CopyProductDataSource dataSource
+    private final CopyProductOutput output
+    private final CreateProductInput createProduct
+
+    /**
+     * The only constructor for this use case
+     * @param dataSource - a data source that provides mandatory functionality
+     * @param output - an output that provides mandatory functionality
+     * @param createProductInput - a CreateProduct use case that is used to create the product
+     */
+    CopyProduct(CopyProductDataSource dataSource, CopyProductOutput output, CreateProductDataSource createProductDataSource) {
+        this.dataSource = dataSource
+        this.output = output
+        this.createProduct = new CreateProduct(createProductDataSource, this)
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    void copyModified(Product product) {
+        Optional<Product> searchResult
+        try{
+            //1. retrieve product from db
+            searchResult = dataSource.fetch(product.getProductId())
+            if (searchResult.isPresent()) {
+                Product existingProduct = searchResult.get()
+                //2. compare if there is a difference between the products in order
+                if(getProductChecksum(product) != getProductChecksum(existingProduct)){
```

Comment:
The OOP way of doing it (and more intuitive and encapsulating) , zhe product checksum should be an intrinsic property of the product and not determined by a method. Then you could have done:
```suggestion
                if(product != existingProduct){
```
way more readable and expressive.